src/box/lua/net_box.c

```diff
@@ -1275,6 +1275,17 @@ static int
 netbox_transport_send_and_recv(struct netbox_transport *transport,
 			       struct xrow_header *hdr)
 {
+	/*
+	 * When the worker fiber enters the Lua VM, garbage collection can be
+	 * triggered and the connection owning the worker could get closed (for
+	 * instance, when decoding the response). Unfortunately, we do not have
+	 * a convenient way to get notified about this event, so the best we can
+	 * do is check the connection state before doing any work.
+	 */
+	if (transport->state == NETBOX_CLOSED) {
+		box_error_raise(ER_NO_CONNECTION, "Connection closed");
+		return -1;
+	}
 	ibuf_consume(&transport->recv_buf, transport->last_msg_size);
 	while (true) {
```

Comment:
Isn't checking `transport->state` at the beginning of the loop step be safer?

Reply:
This is superfluous. The point of this check is to catch the case when we didn't yield, but simply entered the Lua VM and got garbage collected. We don't interact with Lua anywhere inside `netbox_send_and_recv`, and we have separate checks for the case we yield there.